### PR TITLE
Add support for convolutional, pooling and flatten layer

### PR DESCRIPTION
**Pull request by free43**
Added implementation of forward and backward propagation for following layers:
- ConvolutionalLayer
- PoolingLayer
- FlattenLayer

For _calculate_hidden_delta_ calculation is forwarded to the responsible class, so no repeated implementation. 

The used format for input data and their delta is: NxCHxCxR (Batch samples x Channels x Columns x Rows). For kernels the format CHxCxRxK is used. This can be seen in the file [4d_dimensions.h](https://github.com/free43/opennn/blob/master/opennn/4d_dimensions.h).

Added unit and integration test cases. System test is done via the example mnist. Here 70,000 input data is used. Following time measurements are obtained during training on the system OS: Ubuntu 22, Processor: Intel® Core™ i3-2330M CPU @ 2.20GHz × 4, Memory: 8,0 GiB:
- Model 1 without conv layers:  16 s
- Model 2 1xMaxPoolingLayer and 2xconv layers: 36 min 1 s
- Model 3 LeNet5: 30 min 3 s

Further details can be seen in the file [mnist example](https://github.com/free43/opennn/blob/master/examples/mnist/main.cpp). Also the confusion matrix is printed there.

Next steps: performance improvement, adapt other classes like TestingAnalysis to support 3/4D input dimension, add comments and checks

PS: 
 Changed the order of assemble gradient and applying regularization to the weights ([commit](https://github.com/free43/opennn/commit/759abb74d96a31accedbaf78c835b235b0f0876d)). The reason is, assemble gradient will override the applied regularization from the previous batch.

### CODE REVIEW

**Comment by free43** [comment]
> @free43 Is this change adding support in de-convolution as well?

The transposed convolution is implemented, which is often named "deconvolution". See the method ConvolutionalLayer::calculate_hidden_delta. 
The math behind of this can be read here 
[cnn_backpropagation_article.pdf](https://github.com/Artelnics/opennn/files/14824473/cnn_backpropagation_article.pdf)
which was uploaded originally here https://codeandfire.github.io/assets/2021-07-16-backpropagation/article.pdf. Last access 12.10.23. 
The described method can be used to derive a "deconvolution" with multiple kernels and channels.

**Comment by leondavi** [comment]
in batch_normalization_layer.h, display is set to true and there is no setter to change it to false.

@free43 can you please explain the functionality of index parameter to constructor?
explicit BatchNormalizationLayer(const Index&);

**Comment by free43** [comment]
> in batch_normalization_layer.h, display is set to true and there is no setter to change it to false.
> 
> @free43 can you please explain the functionality of index parameter to constructor? explicit BatchNormalizationLayer(const Index&);

I couldn't find the described file or constructor. Could you provide a link to the file?

Besides that, you can create a pull request, where you provide a setter-method for setting the value for display.
